docs/design_doc.md

```diff
@@ -0,0 +1,300 @@
+		# reference to {training,prediction}_q of the ComputationWrapper that uses it
+		self.{training,prediction}_q = {training,prediction}_q
+		# used to accept data returned from ComputationTask's {training,prediction}
+		self.{training,prediction}_return_q = Queue()
+	
+	def put_{training,prediction}_data(self, data):
+		self.{training,prediction}_q.put((self.agent_id, data))
+		
+	def put_{training,prediction}_return(self, data):
+		self.{training,prediction}_return_q.put(data)
+
+	def get_{training,prediction}_return(self):
+		return self.{training,prediction}_q.get()
+```
+
+The definition of `AgentHelper` can then be:
+
+```python
+class AgentHelper(object):
+	def __init__(self):
+		self.name # the name of the task this helper is bound to
+		self.comm # Communicator
+		self.exp_q # a container to store the past experience data
+		
+	def predict(self, inputs):
+		"""
+		send data to ComputationTask for prediction; blocked until return
+		"""
+		inputs = ... # some data processing of inputs
+		self.comm.put_prediction_data(inputs)
+		return self.comm.get_prediction_return()
+		
+	def store_data(self, data):
+		"""
+		store the past experience data, and call learn() if necessary
+		"""
+		self.exp_q.add(data)
+	
+	def learn(self):
+		"""
+		send data to ComputationTask for learning; blocked until return
+		"""
+		# some data preparation, for example:
+		data = self.exp_q.sample()
+		self.comm.put_training_data(data)
+		return self.comm.get_training_return()
+```
+
+Note: depending on the situation, `AgentHelper.learn()` can be called by `Agent` or within `AgentHelper.store_data()`, or be invoked as a separate thread.
+
+As an example, we demonstrate how to define an `Agent` whose robot uses hierarchical policies. The bottom goal-following "RL" policy is updated every `TRAINING_INTERVAL` steps and the top goal generation "goal" policy is updated every time the goal is achieved or aborted.
+
+```python
+class Agent(object):
+	self.env # environment
+	self.helpers = {'RL': OnPolicyHelper, 'goal': OffPolicyHelper} # a dictionary of AgentHelper
+		
+	def run_one_episode()
+		obs = self.env.get_obs()
+		prev_goal = []
+		while not self.env.game_over():
+			goal = self.helpers['goal'].predict([obs, prev_goal])
+			total_reward = 0
+			step_counts = 0
+			while not goal_achieved(goal) and not self.env.game_over():
+				actions = self.helpers['RL'].predict([obs, goal])
+				next_obs, reward = self.env.step(actions)
+				total_reward += reward
+				self.helpers['RL'].store_data([obs, goal, reward, actions, self.env.game_over()])
+				if step_counts % TRAINING_INTERVAL == 0:
+					self.helpers['RL'].learn()
+				obs = next_obs
+			self.helpers['goal'].store_data([obs, goal, total_reward, self.env.game_over()])
+			self.helpers['goal'].learn()
+```
+
+`ComputationWrapper` is essentially two threads that keep consuming data from Agent side:
+
+```python
+class ComputationWrapper(object):                                                
```

Comment:
I don't understand how this Wrapper communicate with simulator.
Are there lots of simulators sending states into self.prediction_q asynchronously?

Reply:
yes. multiple agents (processes) will send data into this q asynchronously